Search in "Manage sets" window

Vafthrudnir opened the pull request:
## Related Ticket(s)
- Fixes #2740 and #2845 

## Short roundup of the initial problem
It wasn't possible to filter or search sets in the `Manage Sets` window before.

## What will change with this Pull Request?
- Search textbox added to filter set selection
- Filtering searches the entered expression in the `Long name`, `Set code`, and `Set type` columns, if either matches the row is shown
- Sorting by a property will not modify the original model order anymore, only explicit drag & drop or moving by button operations are usable for order modification
- If the view has been sorted by clicking on one of the column headers the `Disable sorting` button can be used to retrieve the previous order and re-enable manual order-setting

## Screenshots
![image](https://user-images.githubusercontent.com/9273978/39836796-dcc503a2-53d4-11e8-81a9-43e3c5cd7a05.png)
![image](https://user-images.githubusercontent.com/9273978/39836820-efadd78c-53d4-11e8-8a66-f5e7a2c387a7.png)
![image](https://user-images.githubusercontent.com/9273978/39837023-8584f498-53d5-11e8-90e0-f1e61934a245.png)

ctrlaltca commented on cockatrice/src/window_sets.cpp:
```diff
@@ -173,6 +193,31 @@ void WndSets::actRestore()
 
+void WndSets::actResetSort()
+{
+    view->header()->setSortIndicator(-1, Qt::DescendingOrder);
```
A minor one: you could use a define a explicit name for this magic number, eg.
```
#define SORT_RESET -1
```
and use the name instead of the value.

Vafthrudnir commented:
@ctrlaltca 
- Yeah, I felt like this was not the best name for the button, but nothing else came to mind mind... Though I like this "Restore order", it really makes it clear what it does.
- That workflow won't be bothered by this PR. You still can sort the sets by date, and enable/disable them however you like. You can even filter them by typing "Expansion" to hide any non-standard sets. What won't be possible is rearranging the set order, while the list is sorted, like this:
![sort_and_rearrange](https://user-images.githubusercontent.com/9273978/39839008-288e9e78-53db-11e8-921e-d981c12a330a.gif)

ZeldaZach commented on cockatrice/src/setsmodel.cpp:
```diff
@@ -252,3 +252,37 @@ QStringList SetsModel::mimeTypes() const
+
+bool SetsDisplayModel::filterAcceptsRow(int sourceRow, const QModelIndex &sourceParent) const
+{
+    QModelIndex index0 = sourceModel()->index(sourceRow, SetsModel::SetTypeCol, sourceParent);
```
These names aren't too good...

tooomm commented on cockatrice/src/window_sets.cpp:
```diff
@@ -53,16 +55,32 @@ WndSets::WndSets(QWidget *parent) : QMainWindow(parent)
+    // search field
+    searchField = new QLineEdit;
+    searchField->setObjectName("searchEdit");
+    searchField->setPlaceholderText(tr("Search by card name"));
```
Copy paste error? 😉

Vafthrudnir replied:
Yup, fixed it in next commit 😄

Vafthrudnir commented:
@tooomm 
Thanks for the test! Made the "Enabled" column fixed sized, now it won't change width during filtering. Also added the `(x)` button you requested.

tooomm commented:
After this change, reordering of sets (drag&drop and arrow buttons) is only permitted if you have restored the default order (hitting the button or sort the most recent set to the top). It puzzled me at first why the buttons on the left are disabled and I couldn't enable them when selecting a set...

In general the `Manage sets` dialog a complex window which even gets a bit more complex now, with the very much needed and important search/filter feature.
We should look into decoupling general set management (search + enable) from actual order setting and the connected image load behavior.
People are always confused about that second part and never really understand it at first.
The little explanation we provide at the buttom doesn't help to much with that.
It might be really worth to make that a separate dialog or include it into the settings at one point.

Can you add a explaining hover hint for the restore button? @Vafthrudnir

tooomm commented on cockatrice/src/window_sets.cpp:
```diff
@@ -53,16 +55,33 @@ WndSets::WndSets(QWidget *parent) : QMainWindow(parent)
+    searchField->setClearButtonEnabled(true);
+    setFocusProxy(searchField);
+
+    resetSortButton = new QPushButton(tr("Restore order"));
```
What about `Back to default view`.
That could stress that it's only the displayed view which is going to change, it won't reset your personalized order or restore (=overwrite) changes you made.

It could also show the "mode" you are currently in and highlight that.
`Mode: Picture load ordering` for the disabled button, and `Adjust card art display` for the clickable button.

But that might not yet be thought-out enough... 😄

tooomm commented:
>Could you help me coming up with something useful to show there?

Sure, what about e.g. `Revert displayed set order to your customized picture load priority` or `Show default `?

Vafthrudnir commented:
Thanks for the suggestion @tooomm!
In the latest commit I added the hint text and a new button that restores the original timeline order (new on top).
When the filtering is active (there's something written in the searchbox) this button gets disabled because it wouldn't make much sense to reset set order on only a filtered subset of the sets.
![image](https://user-images.githubusercontent.com/9273978/40204973-a1a6f8be-5a2a-11e8-9dd8-c3d9950acf8e.png)
My concern with this solution is that the function of these two buttons isn't clearly shown by their names. Do you have any suggestions how to solve this issue?

tooomm commented:
Haha, that dialog is a really getting problematic. :D
But making it two won't solve it immediately if you think about about it for a bit.
I still think it would be better most likely. Nothing for this PR though.


>My concern with this solution is that the function of these two buttons isn't clearly shown by their names. Do you have any suggestions how to solve this issue?

Since appveyor is failing currently (they updated two days ago, maybe that's why?), I can't download a test version.

Vafthrudnir commented:
Yeah, I agree :smile: 
Would it be feasible to remove the `Restore order` button, and add that functionality to the column headers? So when a header is clicked, it first orders the sets in descending order, on second click orders it in ascending order, and the third click would restore the original order.
This way a label could pop up somewhere showing why the manual order-changing buttons are disabled, and telling the users how to re-enable them again (click on the column header twice/once). 
Although I can't really see any good substitute for the `Default order` button, something with that functionality should be implemented. What do you think?

Daenyth commented:
> when a header is clicked, it first orders the sets in descending order, on second click orders it in ascending order, and the third click would restore the original order.

If you can do that it would be ideal

Vafthrudnir commented:
Removed the `Restore order` button, and added the same functionality to the column headers (needs to be clicked a third time to disable sorting). When the table is sorted a warning is shown like this:
![image](https://user-images.githubusercontent.com/9273978/40333126-1cf81dc0-5d57-11e8-813c-ff6fe62383ec.png)

tooomm commented:
This is indeed way better and combines both features of the sets dialog quite nicely! Well done.

A few minor things:
- Can you place the warning message right below the list (above the buttons), or below the buttons?
I assume right below the list will feel and look best when it appears, and should be easier to recognize too. Maybe check both and tell us what works best.
Also, having the warning label or all text in red or within a box to make it more obvious is an idea.

- When you sort for a column and hit `Default order` afterwards it looks like everything is back to default (also the sorting), but the warning is still displayed

- It would be nice if hitting `Default order` will bring you back to the top of the list

tooomm commented:
Further observation:

I ran into an issue in combination with automatic spoiler updates / spoiler season:
If a new spoiler file is found you see the "first time user" dialog:
![1](https://user-images.githubusercontent.com/9874850/40494244-ae6ec456-5f74-11e8-8c22-61bd5ddfb2c8.png)

Then, no matter if you click `x` or `OK`, an **empty** "Manage sets..." window opens:
![2](https://user-images.githubusercontent.com/9874850/40494249-b081617c-5f74-11e8-8bd9-741a22c7db93.png)

With the current stable release (2.5.1) I don't even see the "first time user" dialog at all?! 😄

Vafthrudnir commented:
The warning message is definitely more recognizable when it is placed above the buttons, but I think it looks strange there:
![image](https://user-images.githubusercontent.com/9273978/40512606-ac13e036-5fa3-11e8-9fa8-e9a34b7ef6c9.png)
I'd prefer it under the buttons, above the other text:
![image](https://user-images.githubusercontent.com/9273978/40512655-d4e89b82-5fa3-11e8-9fd0-a4d04ff46ca1.png)
What do you think?
Regarding the icons, I think they are missing from the screenshots due to the program running in debug mode, though I'm not sure, there might be something wrong with my environment. Hasn't really bothered me before 😄 

I also checked the manual spoiler update stuff, but I couldn't reproduce the error on my side - it shows the "Spoilers already up to date" message in the bottom right corner, but doesn't show the first time user window. Seems really strange, are you sure this PR causes it?

tooomm commented:
>The warning message is definitely more recognizable when it is placed above the buttons, but I think it looks strange there:

Yeah right. Alternatively the warning could be placed above the set list. Or the buttons could be above and the warning below the list.

>I also checked the manual spoiler update stuff, but I couldn't reproduce the error on my side - it shows the "Spoilers already up to date" message in the bottom right corner, but doesn't show the first time user window. Seems really strange, are you sure this PR causes it?

I  tried again with Client Version: 2.5.2-custom(16e17a2) (2018-05-23), the last commit in this pr. Same result.
If you manually open the manage sets dialog after hitting the update spoiler button in settings and a moment after the notification appears... is the dialog then bugged as well?

skwerlman commented:
You could surround the warning in a box with a slightly different background color to make it stand out where it is

tooomm commented:
Maybe I'm experiencing #3262?

skwerlman commented:
> Maybe I'm experiencing #3264?

What you have screenshotted is definitely #3264

tooomm commented:
>What you have screenshotted is definitely #3264

It should not be caused or related by this PR. :+1:
Weird though, that @Vafthrudnir can't reproduce it.

Vafthrudnir commented:
How about this?
![image](https://user-images.githubusercontent.com/9273978/40871154-8986d98a-6636-11e8-8d79-ad0ef5f7b4fc.png)

tooomm commented:
Can somebody give this a try and double check the code again, please?
